projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+}
+
+// [1] JSON
+func parseJSON(file []byte, dataSlice *[]Player) error {
+	// try to "unmarshal" the data from the json file into the jsonDataSlice variable
+	err := json.Unmarshal(file, &dataSlice)
+
+	// if we fail to "unmarshal" the json file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	return nil
+}
+
+// [2] Repeated JSON
+func parseRepeatedJSON(file []byte, dataSlice *[]Player) error {
+	// convert the file to a string and then split it on a newline character
+	lines := strings.Split(string(file), "\n")
+
+	// if no lines were parsed then error
+	if len(lines) < 1 {
+		return errors.New("[3] could not parse any data in repeated JSON format")
+	}
+
+	// initialise a player struct to iteratively store each valid player
+	var player Player
+
+	// loop over the slice of strings
+	for _, line := range lines {
+
+		// if the line is empty keep going
+		if len(line) == 0 {
+			continue
+		}
+
+		// if the line starts with a # it is a comment so ignore it ((!)use single quotes for characters)
+		if line[0] == '#' {
+			continue
+		}
+
+		// try to parse the player struct from the line and store it in the player struct
+		err := json.Unmarshal([]byte(line), &player)
+
+		// if we can't parse a player struct from the line then error but keep going
+		if err != nil {
+			continue
+		}
+
+		// add the player struct to the slice of players
+		*dataSlice = append(*dataSlice, player)
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in repeated JSON format")
+	}
+
+	return nil
+}
+
+// [3] CSV
+func parseCSV(file []byte, dataSlice *[]Player) error {
+	// because of the way we have tried to use file []byte as a parameter..
+	// it means we have to use bytes.NewReader on the file... this is JANKY
+	reader := csv.NewReader(bytes.NewReader(file))
+
+	// try to read the lines from the csv file
+	lines, err := reader.ReadAll()
+
+	// if we can't read the lines from the csv file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	if len(lines) == 0 {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	// loop over each line
+	for _, line := range lines {
+
+		// (!!!) if we try to read one of the binary files it gives us this: [[��] [AyaPrisha����CharlieMargot]]
+		// so we need to add an additional check here:
+
+		// if the line does not have 2 values (highscore, name) then skip
+		// ((!) this is brittle if a line has valid data but random extra values etc)
+		if len(line) != 2 {
+			continue
+		}
+
+		// get the player name
+		name := line[0]
+
+		// the player has no name so skip
+		if len(name) < 1 {
+			continue
+		}
+
+		//  parse the string to get the score ((!) this is of type int not int32)
```

Comment:
Does this matter? Either way, how could you make an `int32` rather than an `int`?

Reply:
It matters in the sense that the Player struct HighScore is typed as an int32

I think I did make it an int32 in the append later on but that isn't very clean

```
// add the player to the dataSlice
*dataSlice = append(*dataSlice, Player{Name: name, HighScore: int32(highScore)})
```